roachprod: add ui spinners

herkolategan opened the pull request:
Previously, long-running tasks in roachprod had to implement its own logic for
displaying a spinner to indicate progress to the user when running a CLI
command. This logic generally needed to take quiet mode, and file vs. terminal
into account.

This change adds three new reusable implementations: Spinner, CountingSpinner,
GroupSpinner. Spinner is the most basic and renders the spinner when
appropriate. CountingSpinner adds the concept of completed tasks and keeps track
of the count. GroupSpinner is a drop-in replacement for `errgroup.Group` and
adds progress tracking via a CountingSpinner to the group.

Release Note: None
Epic: None

herkolategan committed:
roachprod: add ui spinners

Previously, long-running tasks in roachprod had to implement its own logic for
displaying a spinner to indicate progress to the user when running a CLI
command. This logic generally needed to take quiet mode, and file vs. terminal
into account.

This change adds three new reusable implementations: Spinner, CountingSpinner,
GroupSpinner. Spinner is the most basic and renders the spinner when
appropriate. CountingSpinner adds the concept of completed tasks and keeps track
of the count. GroupSpinner is a drop-in replacement for `errgroup.Group` and
adds progress tracking via a CountingSpinner to the group.

Release Note: None
Epic: None

herkolategan committed:
roachprod: refactor `ParallelE` to use ui spinner

Release Note: None
Epic: None

herkolategan committed:
roachprod: use ui spinner with MIG creation

Release Note: None
Epic: None

herkolategan committed:
roachprod: use ui spinner with LB creation

Release Note: None
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125049)
<!-- Reviewable:end -->

DarrylWong commented on pkg/roachprod/install/cluster_synced.go:
```diff
@@ -2940,36 +2940,16 @@ func (c *SyncedCluster) ParallelE(
 	}
-	defer ticker.Stop()
-
-	var spinner = []string{"|", "/", "-", "\\"}
```
Seeing other usages of the old spinner here:

https://github.com/cockroachdb/cockroach/blob/fbf35e0d1e562dd5372c40b336ae40bad301232e/pkg/roachprod/install/cluster_synced.go#L2184

https://github.com/cockroachdb/cockroach/blob/fbf35e0d1e562dd5372c40b336ae40bad301232e/pkg/roachprod/install/cluster_synced.go#L2572

herkolategan replied:
Good catch! I'll add some commits to cover those as well.

herkolategan replied:
Ah I remember why I skipped these. They are a bit more intricate, may need to be a follow-up PR, but I should at least create an issue that these still need converting.

herkolategan replied:
Adding a new `TaskSpinner` to support `Get` / `Put`. Adds quite a few commits, but will be up today at some point.

DarrylWong commented on pkg/roachprod/ui/spinner.go:
```diff
@@ -0,0 +1,254 @@
+	completed int32
+}
+
+var spinnerLoop = []string{"|", "/", "-", "\\"}
```
`var spinnerLoop = []string{"⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"}`

Do you think this would be too crazy? 😆  (I stole it from the gceworker spinner)

herkolategan replied:
I had the exact same thought! I am 100% for this!

DarrylWong commented on pkg/roachprod/vm/gce/gcloud.go:
```diff
@@ -1982,14 +1985,18 @@ func (p *Provider) CreateLoadBalancer(_ *logger.Logger, vms vm.List, port int) e
-	}
-	cmd := exec.Command("gcloud", args...)
-	output, err := cmd.CombinedOutput()
+	output, err := func() ([]byte, error) {
```
Would adding something like:

```
func (s *Spinner) Run(f func()) {
	defer s.Start()()
	f()
}

----------

var output []byte
ui.NewDefaultSpinner(l, "created health check").Run(func() {
    output, err = cmd.CombinedOutput()
})
```

be slightly cleaner? Although in this you have to declare `output` so I'm not actually sure it's cleaner/less boilerplate.

herkolategan replied:
Yeah I also thought a `Run` function could be a nice to have. Not opposed to adding one, the pain I imagine would be error handling on returns etc would have to be outside.

At this moment I wished there was some way to have a "generic" return pass-through for the `Run`. Alternatively may be good just to provide an error pass-through/wrapper variant of `Run`, then it would just be a bit more nuanced if you needed to return some output as well, which you'll have to capture outside the lambda.

DarrylWong reviewed (approved):
Only one real change I think you should make wrt removing the old spinners, but besides that this looks great!

herkolategan committed:
roachprod: add TaskSpinner

This change adds another ui spinner variant that can display multiple tasks and
update each task line individually.

In order to replace the remaining previous spinner implementation for `Get` and
`Put` roachprod operations, we need to be able to track multiple tasks, some
which may finish earlier than others or even display an error.

Epic: None
Release Note: None

herkolategan committed:
roachprod: use ui spinner in Put

Epic: None
Release Note: None

herkolategan committed:
roachprod: use ui spinner in Get

Epic: None
Release Note: None
